core/src/main/java/com/scalar/db/util/groupcommit/Group.java

```diff
@@ -0,0 +1,242 @@
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// An abstract class that has logics and implementations to manage slots and trigger to emit it.
+abstract class Group<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> {
+  private static final Logger logger = LoggerFactory.getLogger(Group.class);
+
+  protected final Emittable<EMIT_KEY, V> emitter;
+  protected final KeyManipulator<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY> keyManipulator;
+  private final int capacity;
+  private final AtomicReference<Integer> size = new AtomicReference<>();
+  protected final Map<CHILD_KEY, Slot<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V>> slots;
+  // Whether to reject a new value slot.
+  protected final AtomicReference<Status> status = new AtomicReference<>(Status.OPEN);
+
+  // Status of the group.
+  enum Status {
+    // Accepting new slot reservation since the number of slots isn't fixed yet.
+    //
+    // Initial status of groups. No group can move back to OPEN.
+    OPEN(false, false, false),
+
+    // Not accepting new slot reservations since the number of slots is already fixed.
+    // Waiting all the slots are set with values.
+    //
+    // Groups with OPEN status can move to CLOSED.
+    CLOSED(true, false, false),
+
+    // All the slots are set with values. Ready to commit.
+    //
+    // Groups with OPEN or CLOSED status can move to READY.
+    READY(true, true, false),
+
+    // Group commit is done and all the clients have get the results.
+    //
+    // Groups with OPEN, CLOSED or READY status can move to DONE.
+    DONE(true, true, true);
+
+    final boolean isClosed;
+    final boolean isReady;
+    final boolean isDone;
+
+    Status(boolean isClosed, boolean isReady, boolean isDone) {
+      this.isClosed = isClosed;
+      this.isReady = isReady;
+      this.isDone = isDone;
+    }
+  }
+
+  Group(
+      Emittable<EMIT_KEY, V> emitter,
+      KeyManipulator<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY> keyManipulator,
+      int capacity) {
+    this.emitter = emitter;
+    this.keyManipulator = keyManipulator;
+    this.capacity = capacity;
+    this.slots = new HashMap<>(capacity);
+  }
+
+  private boolean noMoreSlot() {
+    return slots.size() >= capacity;
+  }
+
+  abstract FULL_KEY fullKey(CHILD_KEY childKey);
+
+  // If it returns null, the Group is already closed and a retry is needed.
+  @Nullable
+  protected synchronized FULL_KEY reserveNewSlot(
+      Slot<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> slot) {
+    if (isClosed()) {
+      return null;
+    }
+    reserveSlot(slot);
+    if (noMoreSlot()) {
+      close();
+    }
+    return slot.fullKey();
+  }
+
+  private void reserveSlot(Slot<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> slot) {
+    Slot<?, ?, ?, ?, ?> oldSlot = slots.put(slot.key(), slot);
+    if (oldSlot != null) {
+      logger.warn("An old slot exist unexpectedly. {}", oldSlot.fullKey());
```

Comment:
This shouldn't occur, right? If so, just showing the warning message is okay?

Reply:
Good catch! Right. I'll update it to throw AssertionError or IllegalStateException.

Reply:
Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/8d96fd484d03b1d127d338ac3d128ad8d6232bbc